Source/Core/CustomProjection.js

```diff
@@ -0,0 +1,304 @@
+define([
+        './Cartesian3',
+        './Cartographic',
+        './Check',
+        './defaultValue',
+        './defined',
+        './defineProperties',
+        './DeveloperError',
+        './Ellipsoid',
+        './getAbsoluteUri',
+        './loadAndExecuteScript',
+        '../ThirdParty/when'
+    ], function(
+        Cartesian3,
+        Cartographic,
+        Check,
+        defaultValue,
+        defined,
+        defineProperties,
+        DeveloperError,
+        Ellipsoid,
+        getAbsoluteUri,
+        loadAndExecuteScript,
+        when) {
+    'use strict';
+
+    var loadedProjectionFunctions = {};
+
+    /**
+     * {@link MapProjection} that uses custom project and unproject functions defined in an external file.
+     *
+     * The external file must contain a function named <code>createProjectionFunctions</code> that implements the
+     * <code>CustomProjection~factory</code> interface to provide <code>CustomProjection~project</code> and
+     * <code>CustomProjection~unproject</code> functions to a callback.
+     *
+     * Scenes using CustomProjection will default to <code>MapMode2D.ROTATE</code> instead of <code>MapMode2D.INFINITE_SCROLL</code>.
+     *
+     * @alias CustomProjection
+     * @constructor
+     *
+     * @param {String} options.url The url of the external file.
+     * @param {String} options.projectionName A unique name for this projection.
```

Comment:
> Is it possible that name might be unnecessary requirement from the user and that we can use an internal UUID instead? If it needs to be user accessible, it can probably have an id getter. But from what I see, it feels internal only.

It's there for the sake of caching, I didn't want to use the url itself in case the input `url` was a dataURI (as in our Sandcastle examples).

Maybe it should be optional unless the input is a dataURI, in which case we can detect and throw.

Reply:
Why wouldn't generating a UUID be enough for caching? Is there a benefit to having a named identifier?

Reply:
We'd need a way to associate the UUID with the URL that the user is passing in. You're right though, the `projectionName` here isn't very well thought-out. I'll update so we cache by URL directly and just don't cache when the input is a dataURI.